language/stdlib/modules/option.move

```diff
@@ -0,0 +1,119 @@
+module Option {
+    use 0x0::Vector;
+
+    // Abstraction of a value that may or may not be present. Implemented with a vector of size
+    // zero or one because Move bytecode does not have ADTs.
+    struct T<Element> { vec: vector<Element> }
+
+    // Return an empty `Option::T`
+    public fun none<Element>(): T<Element> {
+        T { vec: Vector::empty() }
+    }
+
+    // Return an `Option::T` containing `e`
+    public fun some<Element>(e: Element): T<Element> {
+        T { vec: Vector::singleton(e) }
+    }
+
+    // Return true if `t` does not hold a value
+    public fun is_none<Element>(t: &T<Element>): bool {
+        Vector::is_empty(&t.vec)
+    }
+
+    // Return true if `t` holds a value
+    public fun is_some<Element>(t: &T<Element>): bool {
+        !Vector::is_empty(&t.vec)
+    }
+
+    // Return true if the value in `t` is equal to `e_ref`
+    // Always returns `false` if `t` does not hold a value
+    public fun contains<Element>(t: &T<Element>, e_ref: &Element): bool {
+        Vector::borrow(&t.vec, 0) == e_ref
+    }
+
+    // Return an immutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow<Element>(t: &T<Element>): &Element {
+        Vector::borrow(&t.vec, 0)
+    }
+
+    // TODO: not clear this API is a good idea
+    // Return a reference to the value inside `t` if it holds one
+    // Return `default_ref` if `t` does not hold a value
+    public fun borrow_default<Element>(t: &T<Element>, default_ref: &Element): &Element {
+        let vec_ref = &t.vec;
+        if (Vector::is_empty(vec_ref)) {
+            Vector::borrow(vec_ref, 0)
+        } else {
+            default_ref
+        }
+    }
+
+    // TODO: not clear we want both this and borrow_default
+    // Return the value inside `t` if it holds one
+    // Return `default` if `t` does not hold a value
+    public fun get_default<Element: copyable>(t: &T<Element>, default: Element): Element {
```

Comment:
I don't think you need this, since if you have `Option:: get_default(opt, e)`  you can always rewrite it as `*Option:: borrow_default(opt, &e)`

Reply:
I think this API would be nice if somehow it took an owned version of `Option`? Then at least it feels like it can help in situations where you don't want to copy?

So
```
    public fun get_default<Element: copyable>(t: T<Element>, default: Element): Element {
```

Reply:
> always rewrite it as `*Option:: borrow_default(opt, &e)`

I think it's fair that this isn't so much worse to write, but it does add an extra local, assignment, borrow, and dereference under the hood that can't be optimized away...

Reply:
I'm not convinced those 4 extra operations are a big deal for that specific case. We cannot write library functions that are the most optimal for every possible case. We have to pick ones that the most convenient for the common cases. Thus, I think that `get_default` with a non-reference first value feels better for this operation. 

The copying reference case can be handled with the borrow+dereference snippet

Reply:
> We cannot write library functions that are the most optimal for every possible case. We have to pick ones that the most convenient for the common cases

I agree. I think this function is both convenient and optimal for a common case, however. As weak evidence: the use of `Option` in `ValidatorConfig2` uses this API.